examples/imatrix/imatrix.cpp

```diff
@@ -252,102 +268,136 @@ void IMatrixCollector::save_imatrix(int ncall) const {
             continue;
         }
 
-        n_entries++;
         to_store.push_back(kv.first);
+        data_size += GGML_PAD(ggml_tensor_overhead() + sizeof(float) * kv.second.values.size(), GGML_MEM_ALIGN);
+        data_size += GGML_PAD(ggml_tensor_overhead() + sizeof(float) * kv.second.counts.size(), GGML_MEM_ALIGN);
     }
 
     if (to_store.size() < m_stats.size()) {
         fprintf(stderr, "%s: warning: storing only %zu out of %zu entries\n", __func__, to_store.size(), m_stats.size());
     }
 
-    std::ofstream out(fname, std::ios::binary);
-    out.write((const char *) &n_entries, sizeof(n_entries));
+    // deterministic tensor name order
+    std::sort(to_store.begin(), to_store.end());
+
+    struct ggml_init_params params = {
+        /* .mem_size   = */ data_size,
+        /* .mem_buffer = */ NULL,
+        /* .no_alloc   = */ false,
+    };
+    struct ggml_context * ctx = ggml_init(params);
+    struct gguf_context * ctx_gguf = gguf_init_empty();
+
+    gguf_set_val_str(ctx_gguf, "general.type", "imatrix");
+    // Write the input filename to later on specify it in quantize
+    gguf_set_val_str(ctx_gguf, LLM_KV_IMATRIX_DATASET, m_params.prompt_file.c_str());
+    // Write the number of chunks the matrix was computed with
+    gguf_set_val_u32(ctx_gguf, LLM_KV_IMATRIX_CHUNK_COUNT, m_last_chunk);
+    gguf_set_val_u32(ctx_gguf, LLM_KV_IMATRIX_CHUNK_SIZE, m_params.n_ctx / m_params.n_parallel);
+
     for (const auto & name : to_store) {
         const auto & stat = m_stats.at(name);
-        int len = name.size();
-        out.write((const char *) &len, sizeof(len));
-        out.write(name.c_str(), len);
-        out.write((const char *) &stat.ncall, sizeof(stat.ncall));
-        int nval = stat.values.size();
-        out.write((const char *) &nval, sizeof(nval));
+        const int32_t nval = (int32_t) stat.values.size();
+        const int32_t nmat = (int32_t) stat.counts.size();
         if (nval > 0) {
-            std::vector<float> tmp(nval);
-            for (int i = 0; i < nval; i++) {
-                tmp[i] = (stat.values[i] / static_cast<float>(stat.counts[i])) * static_cast<float>(stat.ncall);
+            struct ggml_tensor * sums = ggml_new_tensor_2d(ctx, GGML_TYPE_F32, nval / nmat, nmat);
+            struct ggml_tensor * counts = ggml_new_tensor_2d(ctx, GGML_TYPE_F32, 1, nmat);
+            ggml_set_name(sums, (name + ".sums").c_str());
+            ggml_set_name(counts, (name + ".counts").c_str());
```

Comment:
There is also `ggml_format_name` that you can use:

```suggestion
            ggml_format_name(sums, "%s.sums", name.c_str());
            ggml_format_name(counts, %s.counts", name.c_str());
```